src/Akeneo/Category/back/Domain/ValueObject/PermissionCollection.php

```diff
@@ -13,80 +13,103 @@ final class PermissionCollection
     public const OWN = 'own';
 
     /** @var array<string, array<int>> */
-    private array $removedPermissions;
+    private array $removedUserGroupIdsFromPermissions;
 
     // @phpstan-ignore-next-line
     private function __construct(private ?array $permissions)
     {
     }
 
     /**
-     * @param array<string, array<int, string>>|null $permissions
+     * @param array<string, array<array{id: int, label: string}>>|null $permissions
      */
     public static function fromArray(?array $permissions): self
     {
         return new self($permissions);
     }
 
     /**
-     * @param array<int> $userGroupIds
+     * @param array<array{id: int, label: string}> $newUserGroups
      */
-    public function addPermission(string $type, array $userGroupIds): self
+    public function addUserGroupsToPermission(string $permission, array $newUserGroups): self
     {
-        if (array_key_exists($type, $this->permissions)) {
-            foreach ($userGroupIds as $userGroupId) {
-                if (!in_array($userGroupId, $this->permissions[$type])) {
-                    $this->permissions[$type][] = $userGroupId;
+        if (array_key_exists($permission, $this->permissions)) {
+            $existingUserGroupsIds = array_map(fn ($existingUserGroup) => $existingUserGroup['id'], $this->permissions[$permission]);
+
+            foreach ($newUserGroups as $newUserGroup) {
+                if (!in_array($newUserGroup['id'], $existingUserGroupsIds)) {
+                    $this->permissions[$permission][] = $newUserGroup;
                 }
             }
         } else {
-            $this->permissions[$type] = $userGroupIds;
+            $this->permissions[$permission] = $newUserGroups;
         }
 
         return new self($this->permissions);
     }
 
     /**
-     * @param array<int> $userGroupIds
+     * @param array<array{id: int, label: string}> $userGroupsToRemove
      */
-    public function removePermission(string $type, array $userGroupIds): self
+    public function removeUserGroupsFromPermission(string $permission, array $userGroupsToRemove): self
     {
-        if (array_key_exists($type, $this->permissions)) {
-            foreach ($userGroupIds as $userGroupId) {
-                if (($key = array_search($userGroupId, $this->permissions[$type])) !== false) {
-                    $this->removedPermissions[$type][] = $userGroupId;
-                    unset($this->permissions[$type][$key]);
+        if (array_key_exists($permission, $this->permissions)) {
+            $existingUserGroupsIds = array_map(fn ($existingUserGroup) => $existingUserGroup['id'], $this->permissions[$permission]);
```

Comment:
Just a little refacto. You use thiis map function sevrale times.
```suggestion
            $existingUserGroupsIds = $this->getUserGroupIds();
```